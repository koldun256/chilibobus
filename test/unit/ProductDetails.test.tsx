import React, { act } from "react";
import { ProductDetails } from "../../src/client/components/ProductDetails";
import { render, screen } from "@testing-library/react";
import { addToCart, initStore } from "../../src/client/store";
import { BrowserRouter } from "react-router-dom";
import "@testing-library/jest-dom";
import { Provider } from "react-redux";
import { CartApi, ExampleApi } from "../../src/client/api";
import userEvent from "@testing-library/user-event";

const product = {
  id: 69,
  name: "bipki",
  price: 420,
  description: "lorem",
  material: "ipsum",
  color: "doret",
};
function renderProductDetails() {
  localStorage.clear();
  const api = new ExampleApi("/hw/store");
  const cart = new CartApi();
  const store = initStore(api, cart);
  const { container } = render(
    <BrowserRouter>
      <Provider store={store}>
        <ProductDetails product={product} />
      </Provider>
    </BrowserRouter>
  );
  return { store, container };
}
describe("Страница с подробностями о товаре", () => {
  it("показывает информацию о товаре", () => {
    const { container } = renderProductDetails();

    expect(container).toMatchSnapshot("ProductDetails без корзины");
  });
  it("при добавлении в корзину отображает соответствующий индикатор", () => {
    const { container, store } = renderProductDetails();

    act(() => store.dispatch(addToCart(product)));

    expect(container).toMatchSnapshot("ProductDetails с корзиной");
  });
  it("добавляет товар в корзину при клике на кнопку", async () => {
    const { store } = renderProductDetails();
    const user = userEvent.setup();

    await user.click(screen.getByText(/add to cart/i));
    await user.click(screen.getByText(/add to cart/i));

    expect(store.getState().cart).toEqual({
      [product.id]: { name: product.name, count: 2, price: product.price },
    });
  });
});
