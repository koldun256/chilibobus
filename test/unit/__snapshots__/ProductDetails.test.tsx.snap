// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Страница с подробностями о товаре показывает информацию о товаре: ProductDetails без корзины 1`] = `
<div>
  <div
    class="ProductDetails row"
  >
    <div
      class="col-12 col-sm-5 col-lg-4"
    >
      <img
        class="Image"
        src="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAAEAAAABCAQAAAC1HAwCAAAAC0lEQVR42mNkMAYAADkANVKH3ScAAAAASUVORK5CYII="
      />
    </div>
    <div
      class="col-12 col-sm-7 col-lg-6"
    >
      <h1
        class="ProductDetails-Name"
      >
        bipki
      </h1>
      <p
        class="ProductDetails-Description"
      >
        lorem
      </p>
      <p
        class="ProductDetails-Price fs-3"
      >
        $
        420
      </p>
      <p>
        <button
          class="ProductDetails-AddToCart btn btn-primary btn-lg"
        >
          Add to Cart
        </button>
      </p>
      <dl>
        <dt>
          Color
        </dt>
        <dd
          class="ProductDetails-Color text-capitalize"
        >
          doret
        </dd>
        <dt>
          Material
        </dt>
        <dd
          class="ProductDetails-Material text-capitalize"
        >
          ipsum
        </dd>
      </dl>
    </div>
  </div>
</div>
`;

exports[`Страница с подробностями о товаре при добавлении в корзину отображает соответствующий индикатор: ProductDetails с корзиной 1`] = `
<div>
  <div
    class="ProductDetails row"
  >
    <div
      class="col-12 col-sm-5 col-lg-4"
    >
      <img
        class="Image"
        src="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAAEAAAABCAQAAAC1HAwCAAAAC0lEQVR42mNkMAYAADkANVKH3ScAAAAASUVORK5CYII="
      />
    </div>
    <div
      class="col-12 col-sm-7 col-lg-6"
    >
      <h1
        class="ProductDetails-Name"
      >
        bipki
      </h1>
      <p
        class="ProductDetails-Description"
      >
        lorem
      </p>
      <p
        class="ProductDetails-Price fs-3"
      >
        $
        420
      </p>
      <p>
        <button
          class="ProductDetails-AddToCart btn btn-primary btn-lg"
        >
          Add to Cart
        </button>
        <span
          class="CartBadge text-success mx-3"
        >
          Item in cart
        </span>
      </p>
      <dl>
        <dt>
          Color
        </dt>
        <dd
          class="ProductDetails-Color text-capitalize"
        >
          doret
        </dd>
        <dt>
          Material
        </dt>
        <dd
          class="ProductDetails-Material text-capitalize"
        >
          ipsum
        </dd>
      </dl>
    </div>
  </div>
</div>
`;
